File: laythe_vm/fixture/rb_benchmark/method_call.rb
class Toggle
  def initialize(startState)
    @state = startState
  end

  def value
    @state
  end

  def activate
    @state = !@state
    self
  end
end

class NthToggle < Toggle
  def initialize(startState, maxCounter)
    super(startState)
    @countMax = maxCounter
    @count = 0
  end

  def activate
    @count += 1
    if @count >= @countMax
      super()
      @count = 0
    end

    self
  end
end

start = Time.new
n = 100000
val = true
toggle = Toggle.new(val)

n.times do |_|
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
end

puts toggle.value()

val = true
ntoggle = NthToggle.new(val, 3)

n.times do |_|
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
end

puts ntoggle.value
puts Time.new - start

File: laythe_vm/fixture/py_benchmark/method_call.py
import time

class Toggle:
  def __init__(self, startState):
    self.state = startState

  def value(self):
    return self.state

  def activate(self):
    self.state = not self.state
    return self

class NthToggle(Toggle):
  def __init__(self, startState, maxCounter):
    super().__init__(startState)
    self.countMax = maxCounter
    self.count = 0

  def activate(self):
    self.count += 1
    if self.count >= self.countMax:
      super().activate()
      self.count = 0

    return self

start = time.time()
n = 100000
val = True
toggle = Toggle(val)

for _ in range(n):
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()
  val = toggle.activate().value()

print(toggle.value())

val = True
ntoggle = NthToggle(val, 3)

for _ in range(n):
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()
  val = ntoggle.activate().value()

print(ntoggle.value())
print(time.time() - start)
